src/server/replica.cc

```diff
@@ -1152,29 +1141,73 @@ error_code Replica::SendCommand(string_view command, ReqSerializer* serializer)
   return ec;
 }
 
-bool Replica::TranactionData::UpdateFromParsedEntry(journal::ParsedEntry&& entry) {
-  if (entry.opcode == journal::Op::EXEC) {
-    shard_cnt = entry.shard_cnt;
-    dbid = entry.dbid;
-    txid = entry.txid;
-    return true;
-  } else if (entry.opcode == journal::Op::COMMAND) {
-    txid = entry.txid;
-    shard_cnt = entry.shard_cnt;
-    dbid = entry.dbid;
-    commands.push_back(std::move(entry.cmd));
-    return true;
-  } else if (entry.opcode == journal::Op::MULTI_COMMAND) {
-    commands.push_back(std::move(entry.cmd));
-    return false;
-  } else {
-    DCHECK(false) << "Unsupported opcode";
+bool Replica::TransactionData::AddEntry(journal::ParsedEntry&& entry) {
+  switch (entry.opcode) {
+    case journal::Op::COMMAND:
+    case journal::Op::EXPIRED:
+      commands.push_back(std::move(entry.cmd));
+      [[fallthrough]];
+    case journal::Op::EXEC:
+      shard_cnt = entry.shard_cnt;
+      dbid = entry.dbid;
+      txid = entry.txid;
+      return true;
+    case journal::Op::MULTI_COMMAND:
+      commands.push_back(std::move(entry.cmd));
+      return false;
+    default:
+      DCHECK(false) << "Unsupported opcode";
   }
   return false;
 }
 
-bool Replica::TranactionData::IsGlobalCmd() const {
+bool Replica::TransactionData::IsGlobalCmd() const {
   return commands.size() == 1 && commands.front().cmd_args.size() == 1;
 }
 
+auto Replica::TransactionReader::Next(JournalReader* reader, Context* cntx)
+    -> optional<TransactionData> {
+  io::Result<journal::ParsedEntry> res;
+  do {
+    if (res = reader->ReadEntry(); !res) {
+      cntx->ReportError(res.error());
+      return std::nullopt;
+    }
+
+    // Expiration is preceding to a MULTI...EXEC sequence, even if it was issued within it.
+    if (res->opcode == journal::Op::EXPIRED) {
+      TransactionData tmp_tx;
+      CHECK(tmp_tx.AddEntry(std::move(*res)));
+      return tmp_tx;
+    }
+
+  } while (!cntx->IsCancelled() && !current_.AddEntry(std::move(*res)));
+
+  return cntx->IsCancelled() ? std::nullopt : make_optional(std::move(current_));
+}
```

Comment:
But the whole reason why I needed to modify the replica and why I introduced a stateful TransactionReader, is because I need to catch expirations that are triggered inside a MULTI-EXEC sequence and execute them before